### PR TITLE
Modernize portfolio optimizer

### PR DESCRIPTION
**Pull request by jtao1**
# Modernizing an example checklist

## Preliminary checks

- [x] Look for [open PRs](https://github.com/holoviz-topics/examples/pulls) and [issues](https://github.com/holoviz-topics/examples/issues) that reference the project you are updating. It is possible previous unmerged work in PR could be re-used to modernize the project. Comment on these PRs and issues when appropriate, hopefully we should be able to close some of them after your modernizing work.

## Change ‘anaconda-project.yml’ to use the latest workable version of packages

- [x] Pin python=3.11
- [x] Remove the upper pin (e.g. `hvplot<0.9` to `hvplot`, `panel>=0.12,<1.0` to `panel>=0.12`) of all other dependencies. Removing the upper pins of dependencies could necessitate code revisions in the notebooks to address any errors encountered in the updated environment. Should complexities or extensive time requirements arise, document issues for team discussion on whether to re-pin specific packages or explore other solutions.
- [x] Add/update the lower pin of all other dependencies (e.g. `hvplot` to `hvplot>=0.9.2`, `hvplot>=0.8` to `hvplot>=0.9.2`). Usually, the new/updated lower pin of a dependency will be the version resolved after `anaconda prepare` has been run. Execute `!conda list` in a notebook, or `anaconda run conda list` in the terminal, to display the version of each dependency installed in the environment. Adjusting the lower pin helps ensure that the locks produced for each platform (linux-64, win-64, osx-64, osx-arm64) rely on the tested dependencies and not on some older versions.
- [x] If one of the channels include conda-forge or pyviz, ask Maxime if it can be removed

## Plot API updates (discussed on a per-example basis)

- [x] Generally, try to replace HoloViews usage with hvPlot. At a certain point of complexity, such as with the use of ‘.select’, it might be better to stick with HoloViews. Additional examples of ‘complexity boundaries’ should be documented in [this document](https://docs.google.com/document/d/1LYkChbVZzqq5S1S4Mkj1rM3PNPJDDGriks8YTc4EQc4/edit?usp=sharing).
- [x] Almost always, try to replace the use of `datashade` with `rasterize` (read [this page](https://holoviews.org/user_guide/Large_Data.html)). Essentially, `rasterize` allows Bokeh to handle the colormapping instead of Datashader.

## Interactivity API updates (discussed on a per-example basis)

- [x] Remove all `pn.interact` usage
- [x] Avoid `.param.watch()` usage. This is [pretty low-level and verbose approach](https://param.holoviz.org/user_guide/Dependencies_and_Watchers.html#watchers) and should not be used in Examples unless required, or an Example is specifically trying to demo its usage in an advanced workflow.
- [x] Prefer using `pn.bind()`. Read [this page](https://panel.holoviz.org/explanation/api/reactive.html) for explanation.
- [x] For apps built using a class approach, when they create a `view()` method and call it directly, update the class by inheriting from `pn.viewable.Viewer` and replace `view()` by `__panel__()`. Here is an [example](https://panel.holoviz.org/gallery/nyc_deckgl.html). 

## Panel App updates (discussed on a per-example basis)

- [x] If the project doesn’t at any point create a Panel app at all, consider creating one. It can be as simple as wrapping a plot in `pn.Column`, or more complicated to incorporate widgets, etc. Make the final app  `.servable()`.
- [x] If the project creates an app in a notebook but doesn’t deploy it (i.e. there is no `command: dashboard` declaration in the `anaconda-project.yml` file), try adding it.
- [x] If the project already deploys an app but doesn’t wrap it in a nice template, consider wrapping it in a [template](https://panel.holoviz.org/reference/index.html#templates).
- [x] If the project deploys an app wrapped in a template, customize the template a little so all the apps don’t look similar (e.g. change the header background color). This doesn’t need to be discussed.
- [x] Comment start If you are building the application in a single cell, you can construct a template explicitly, like `template = pn.template.BootstrampTemplate`, but if building up an app across multiple cells, it is probably cleaner to declare the template at the top with `pn.extension(template='bootstrap')`. See [how to guide on setting a template](https://panel.holoviz.org/how_to/templates/template_set.html).

## General code quality updates

- [x] If the notebook disables warnings (e.g. with `warnings.simplefilter(‘ignore’)` somewhere at the start of the notebook, remove this line. Try to update the code to remove the warnings, if any. If updating the code to remove the warnings is taking significant amount of time and effort, bring it up for discussion and we may decide to disable warnings again.

## Text content

- [x] Edit the text content anywhere and everywhere that it can be improved for clarity.
- [x] Check the links are valid, and update old links (e.g. http -> https, xyz.pyviz.org -> xyz.holoviz.org)
- [x] Remove instructions to install packages inside an example

## Visual appearance - Example

- [x] Check that the titles/headings make sense and are succinct.
- [x] Check that the text content blocks are easily readable; revise into additional paragraphs if needed.
- [x] Check that the code blocks are easily readable; revise as needed. (e.g. add spaces after commas in a list if there are none, wrap long lines, etc.)
- [x] Check image and plot sizes. If possible, making them responsive is highly recommended.
- [x] Check the appearance on a smartphone (check Google to see how to adapt the appearance of your browser to display pages as if they were seen from a smartphone, this is usually done via the web developer tools). This is not a top priority for all examples, but if there are a few easy and straightforward changes to make that can improve the experience, let’s do it.
- [x] Check the updated notebook with the original notebook

## Visual appearance - Gallery

- [x] Check the thumbnail is visually appealing
- [x] Check the project title is well formatted (e.g. `Ml Annotators` to `ML Annotators`), if not, add/update the `examples_config.title` field in `anaconda-project.yml`
- [x] Check the project description is appropriate, if not, update the `description` field in `anaconda-project.yml`

## Workflow (after you have made the changes above)

- [x] Run successfully `doit validate:<projectname>`
- [x] Run successfully `doit test:<projectname>`
- [x] Run successfully `doit doc_one –name <projectname>`. It’s better if the project notebook(s) is saved with its outputs (but be sure to clear outputs before committing to the examples repo!) when building the docs. Then open this file in your browser `./builtdocs/index.html` and check how the site looks.
- [x] If you’re happy with all the above, open a PR. Reminder, clear notebook outputs before pushing to the PR.

### CODE REVIEW

**Review by maximlt — CHANGES_REQUESTED** [review]
I have reviewed the PR. Before adding general comments, I wanted to ask you the status of these tasks you wanted to complete?

> Submitted first update on portfolio optimizer. Some things that I still need to do:
> 
>     * added better explanations for the code especially the newer ones related to creating the panel app
> 
>     * remove possibly redundant math when creating the graphs

Comments after review:

- It took me a little while to get what `252` stood for, it turns out it's the number of trading days in a year. Make this a little clearer in the notebook please.
- The first part of the example, that was already there, is quite dry. I don't think spending time improving that is really in the scope of our work, unless that's something that would be quick and easy for you to do!
- However, I think the introduction could be improved. At the moment it only talks about the financial aspect of the example, it could be extended a little to describe the HoloViz tools that are going to be used (hvplot, holoviews, panel) and how they support the analysis. It should also be mentioned a Panel dashboard is going to be created. Adding a screenshot of this app at the beginning would be nice (check out Isaiah's PR on `nyc_taxi` and the start of the `dashboard.ipynb` notebook).
- My main feedback: the section you have ported from Panel's website is also very dry. We need to add some explanation about the app we are building and how we are building it. The `Build a <app>` tutorials you have recently gone through could be a good template for this.
- We should add this app as a deployment (check out `./template/anaconda-project.yaml` to see how). And if we do that, we could also wrap it in a template at the very end of the notebook, so the deployed app looks good.
- Some of the elements of the app could be made responsive (i.e. set up so they dynamically stretch to occupy the available space).
- The app is built using hvPlot's interactive API. It's possible we'll try to migrate to `pn.rx`. I'm waiting for feedback from others on Discord on this point.

EDIT: I updated your original post with the modernizing checklist.

**Comment by maximlt** [comment]
I have a suggestion about the notebook being a little dry and lacking context. Maybe you could throw the notebook content at ChatGPT and ask for it to add some brief explanations about the underlying analysis. That's just a suggestion, so feel free to ignore it if you don't feel like it. I think though that ChatGPT could do a good job at this, given this is I believe a very common kind of analysis.

**Comment by jbednar** [comment]
> It's possible we'll try to migrate to pn.rx

Is there anyone who is arguing for keeping it hvplot.interactive? pn.rx seems like the obvious thing to do, unless someone is very convincingly arguing otherwise.

File: portfolio_optimizer/portfolio_optimizer.ipynb [hunk 1]
```diff
@@ -34,9 +34,11 @@
    "metadata": {},
    "outputs": [],
    "source": [
+    "from io import BytesIO\n",
     "import numpy as np\n",
     "import pandas as pd\n",
-    "import hvplot.pandas  # noqa"
+    "import hvplot.pandas  # noqa\n",
+    "import panel as pn"
```
[the remark this answers is not in the document]
Comment (jtao1) [reply portfolio_optimizer/portfolio_optimizer.ipynb:44]: This has been merged.

**Comment by ahuang11** [comment]
Was helping Jason with `rx` and it felt like `rx` was a bit magical / unintuitive to me.

Also, we weren't sure how to separate out the widgets to better lay out the dashboard, unlike pn.Param.

I'm -1 on using `rx` and prefer `pn.bind` way more.

```python
from io import BytesIO
import pandas as pd
import panel as pn
pn.extension()

@pn.cache
def get_stocks(data):
    if data is None:
        stock_file = 'https://datasets.holoviz.org/stocks/v1/stocks.csv'
    else:
        stock_file = BytesIO(data)
    return pd.read_csv(stock_file, index_col='Date', parse_dates=True)


file_input = pn.widgets.FileInput(sizing_mode='stretch_width')

stocks = pn.rx(get_stocks)(file_input)

selector = pn.widgets.MultiSelect(
    name='Select stocks', sizing_mode='stretch_width',
    options=stocks.columns.to_list()
)

selected_stocks = stocks.rx.pipe(
    lambda df, cols: df[cols] if cols else df, selector
)

pn.Column(selected_stocks)
```
image

Lastly, upon printing rx objects, it raises an error
https://github.com/holoviz/param/issues/939

**Comment by ahuang11** [comment]
Ok I just realized there's a [section in Panel about rx](https://panel.holoviz.org/tutorials/basic/pn_rx.html); I was too focused on the param docs.

To split the widget + column, you need to wrap it in the corresponding pane; personally I'd prefer a method to call to do so, e.g. `selected_stocks.get_pane()`

```
from io import BytesIO
import pandas as pd
import panel as pn
pn.extension()

@pn.cache
def get_stocks(data):
    if data is None:
        stock_file = 'https://datasets.holoviz.org/stocks/v1/stocks.csv'
    else:
        stock_file = BytesIO(data)
    return pd.read_csv(stock_file, index_col='Date', parse_dates=True)


file_input = pn.widgets.FileInput(sizing_mode='stretch_width')

stocks = pn.rx(get_stocks)(file_input)

selector = pn.widgets.MultiSelect(
    name='Select stocks', sizing_mode='stretch_width',
    options=stocks.columns.to_list()
)

selected_stocks = stocks.rx.pipe(
    lambda df, cols: df[cols] if cols else df, selector
)

pn.Column(
    pn.Row(file_input, selector),
    pn.pane.DataFrame(selected_stocks),
)
```

File: portfolio_optimizer/portfolio_optimizer.ipynb [hunk 2]
```diff
@@ -140,7 +143,8 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "log_ret.hvplot.hist(bins=100, subplots=True, width=400, group_label='Ticker', grid=True).cols(2)"
+    "log_ret_hists = log_ret.hvplot.hist(bins=100, subplots=True, width=400, group_label='Ticker', grid=True).cols(2)\n",
```
[the remark this answers is not in the document]
Comment (jtao1) [reply portfolio_optimizer/portfolio_optimizer.ipynb]: I've made the change for spreading the code over two lines.

**Review by maximlt — DISMISSED** [review]

**Comment by jbednar** [comment]
> @ahuang11 : Ok I just realized there's a [section in Panel about rx](https://panel.holoviz.org/tutorials/basic/pn_rx.html); I was too focused on the param docs.

Can you please make a PR on Param putting in an explicit link to the Panel rx docs wherever you would have found them? Right now it only links to the main Panel site, but pointing directly to Panel rx info would presumably prevent others from having the same experience you did.

File: portfolio_optimizer/portfolio_optimizer.ipynb [hunk 3]
```diff
@@ -331,11 +343,16 @@
...
+    "data = pd.DataFrame({\n",
+    "    'Volatility': vol_arr,\n",
+    "    'Return': ret_arr,\n",
+    "    'Sharpe Ratio': sharpe_arr\n",
+    "})\n",
+    "max_sharpe_data = pd.DataFrame({'vol': [max_sr_vol], 'ret': [max_sr_ret]})\n",
     "\n",
-    "scatter.opts(color='Sharpe Ratio', cmap='plasma', width=600, height=400, colorbar=True, padding=0.1) *\\\n",
-    "max_sharpe.opts(color='red', line_color='black', size=10)"
+    "scatter = data.hvplot.scatter(x='Volatility', y='Return', c='Sharpe Ratio', cmap='plasma', width=600, height=400, colorbar=True, padding=0.1)\n",
```
[the remark this answers is not in the document]
Comment (jtao1) [reply portfolio_optimizer/portfolio_optimizer.ipynb:353]: I reduced the marker size to 25, I thought 10 looked too small.

File: portfolio_optimizer/portfolio_optimizer.ipynb [hunk 4]
```diff
@@ -548,7 +565,260 @@
...
+    "    if data is None:\n",
+    "        stock_file = 'https://datasets.holoviz.org/stocks/v1/stocks.csv'\n",
+    "    else:\n",
+    "        stock_file = BytesIO(data)\n",
+    "    return pd.read_csv(stock_file, index_col='Date', parse_dates=True)\n",
+    "\n",
+    "\n",
+    "file_input = pn.widgets.FileInput(sizing_mode='stretch_width')\n",
+    "\n",
+    "stocks = hvplot.bind(get_stocks, file_input).interactive()\n",
```
[the remark this answers is not in the document]
Comment (jtao1) [reply portfolio_optimizer/portfolio_optimizer.ipynb:658]: Done!

File: portfolio_optimizer/portfolio_optimizer.ipynb [hunk 5]
```diff
@@ -548,7 +565,260 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import holoviews as hv\n",
+    "pn.extension('tabulator', design='material', template='material', theme_toggle=True, loading_indicator=True)\n",
```
[the remark this answers is not in the document]
Comment (jtao1) [reply portfolio_optimizer/portfolio_optimizer.ipynb:587]: Done!

File: portfolio_optimizer/portfolio_optimizer.ipynb
[the remark this answers is not in the document]
Comment (jtao1) [reply portfolio_optimizer/portfolio_optimizer.ipynb:1]: Done

File: portfolio_optimizer/portfolio_optimizer.ipynb [hunk 6]
```diff
@@ -11,11 +11,14 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Portfolio Optimization is used for risk-averse investors to construct portfolios to optimize or maximize expected return based on a given level of market risk, emphasizing that risk is an inherent part of higher reward\n",
+    "Portfolio Optimization helps risk-averse investors construct portfolios to maximize expected return given a certain level of market risk, emphasizing the inherent relationship between risk and reward. We will begin by running an example of the Monte Carlo Simulation for an optimal portfolio with resulting returns. Then we will create an Efficient Frontier which is used to identify a set of optimal portfolios that offers the highest expected return for a defined level of risk or the lowest risk for a given level of expected return. This will be done using hvPlot and HoloViz which will allow users to visualize the trade-offs between risk and return. Lastly, we will combine all our analysis into a Panel app that enables users to dynamically explore the Efficient Frontier, adjust parameters, and visualize the resulting portfolios, streamlining the portfolio optimization process.\n",
```
[the remark this answers is not in the document]
Comment (jtao1) [reply portfolio_optimizer/portfolio_optimizer.ipynb:14]: Oh yep, your way makes the sentence clearer.

**Comment by droumis** [comment]
I made a lot of changes. This PR wasn't really close to being ready for a final review.

File: portfolio_optimizer/portfolio_optimizer.ipynb
[the remark this answers is not in the document]
Comment (droumis) [reply portfolio_optimizer/portfolio_optimizer.ipynb:1]: there were still heading errors. I think I caught the rest

**Review by droumis — APPROVED** [review]

**Pull request merged**